dlt/destinations/impl/clickhouse/clickhouse.py

```diff
@@ -0,0 +1,277 @@
+    "merge_tree": "MergeTree",
+    "replicated_merge_tree": "ReplicatedMergeTree",
+}
+
+
+class ClickhouseTypeMapper(TypeMapper):
+    sct_to_unbound_dbt = {
+        "complex": "String",
+        "text": "String",
+        "double": "Float64",
+        "bool": "Boolean",
+        "date": "Date",
+        "timestamp": "DateTime('UTC')",
+        "time": "Time('UTC')",
+        "bigint": "Int64",
+        "binary": "String",
+        "wei": "Decimal",
+    }
+
+    sct_to_dbt = {
+        "decimal": "Decimal(%i,%i)",
+        "wei": "Decimal(%i,%i)",
+        "timestamp": "DateTime(%i, 'UTC')",
+        "time": "Time(%i ,'UTC')",
+    }
+
+    dbt_to_sct = {
+        "String": "text",
+        "Float64": "double",
+        "Boolean": "bool",
+        "Date": "date",
+        "DateTime": "timestamp",
+        "DateTime('UTC')": "timestamp",
+        "Time": "timestamp",
+        "Time('UTC')": "timestamp",
+        "Int64": "bigint",
+        "JSON": "complex",
+        "Decimal": "decimal",
+    }
+
+    def to_db_time_type(self, precision: Optional[int], table_format: TTableFormat = None) -> str:
+        return "DateTime"
+
+    def from_db_type(
+        self, db_type: str, precision: Optional[int] = None, scale: Optional[int] = None
+    ) -> TColumnType:
+        if db_type == "Decimal" and (precision, scale) == self.capabilities.wei_precision:
+            return dict(data_type="wei")
+        return super().from_db_type(db_type, precision, scale)
+
+
+class ClickhouseLoadJob(LoadJob, FollowupJob):
+    def __init__(
+        self,
+        file_path: str,
+        table_name: str,
+        client: ClickhouseSqlClient,
+        staging_credentials: Optional[CredentialsConfiguration] = None,
+    ) -> None:
+        file_name = FileStorage.get_file_name_from_file_path(file_path)
+        super().__init__(file_name)
+
+        qualified_table_name = client.make_qualified_table_name(table_name)
+
+        bucket_path: str = (
+            NewReferenceJob.resolve_reference(file_path)
+            if NewReferenceJob.is_reference_job(file_path)
+            else ""
+        )
+        file_name = (
+            FileStorage.get_file_name_from_file_path(bucket_path) if bucket_path else file_name
+        )
+        file_extension = os.path.splitext(file_name)[1].lower()
+        if file_extension not in ["parquet", "jsonl"]:
+            raise ValueError("Clickhouse staging only supports 'parquet' and 'jsonl' file formats.")
+
+        if not bucket_path:
+            # Local filesystem.
+            raise NotImplementedError("Only object storage is supported.")
+
+        bucket_url = urlparse(bucket_path)
+        bucket_scheme = bucket_url.scheme
+
+        table_function: str
+
+        if bucket_scheme in ("s3", "gs", "gcs"):
+            bucket_http_url = convert_storage_to_http_scheme(bucket_url)
+
+            table_function = (
+                render_s3_table_function(
```

Comment:
so this s3 rendering also works for gcs? maybe then rename the function, but from the code in there I'm not sure it actually works this way?

Reply:
From https://clickhouse.com/docs/en/sql-reference/table-functions/gcs:

> This is an alias of the [s3 table function](https://clickhouse.com/docs/en/sql-reference/table-functions/s3).

The s3 table function should work for both **s3** and **gcs**, but not for **az**.  I've renamed the function to `render_object_storage_table_function` for clarity. This function should ideally contain azure logic as well.